File: ROCK-PAPER-SCISSORS-LIZARD-SPOCK/game.rb
players = [
  {:num=>133, :sign=>"P", wins:[]},
  {:num=>284, :sign=>"R", wins:[]},
  {:num=>420, :sign=>"L", wins:[]},
  {:num=>229, :sign=>"C", wins:[]},
  {:num=>805, :sign=>"L", wins:[]},
  {:num=>262, :sign=>"R", wins:[]},
  {:num=>34, :sign=>"R", wins:[]},
  {:num=>45, :sign=>"L", wins:[]}
]

def round_same_sign(player1, player2)
  winner = player1[:num] < player2[:num] ? player1 : player2
  winner == player1 ? winner[:wins] << player2[:num] :  winner[:wins] << player1[:num]
  return winner
end

def round(player1, player2)
  sign1 = player1[:sign]
  sign2 = player2[:sign]
  case sign1
  when 'R'
    winner = %w[P S].include?(sign2) ? player2 : player1
  when 'P'
    winner = %w[C L].include?(sign2) ? player2 : player1
  when 'C'
    winner = %w[R S].include?(sign2) ? player2 : player1
  when 'L'
    winner = %w[R C].include?(sign2) ? player2 : player1
  when 'S'
    winner = %w[L P].include?(sign2) ? player2 : player1
  end
  winner == player1 ? winner[:wins] << player2[:num] :  winner[:wins] << player1[:num]
  winner
end

# p round( {:num=>420, :sign=>"L", wins:[]},
#   {:num=>229, :sign=>"C", wins:[]})

# Parcourir players tant que players.length > 1
while players.length > 1
  # Instantier winners à liste vide
  winners = []
  l = players.length
  (0..(l / 2 - 1)).to_a.each do |index|
    player1 = players[index * 2]
    player2 = players[index * 2 + 1]
    # Pour chaque paire de joueurs
    # Méthode qui décide du gagnant pour 2 joueurs et qui retourne le joueur gagnant
    if player1[:sign] == player2[:sign]
      winner = round_same_sign(player1, player2)
    else
      winner = round(player1, player2)
    end
    # Pusher le winner dans winners
    winners << winner
  end
  # Lorsque fini de parcourir players, remplacer players par winners
  players = winners
end

# p players
puts players.first[:num]
puts players.first[:wins].join(' ')

File: ROCK-PAPER-SCISSORS-LIZARD-SPOCK/game.ts
type Sign = "R" | "P" | "C" | "L" | "S"

type Player = {
  num: number,
  sign: Sign,
  wins: number[]
}

let players: Player[] = [
  {num: 133, sign: "P", wins:[]},
  {num: 284, sign: "R", wins:[]},
  {num: 420, sign: "L", wins:[]},
  {num: 229, sign: "C", wins:[]},
  {num: 805, sign: "L", wins:[]},
  {num: 262, sign: "R", wins:[]},
  {num: 34, sign: "R", wins:[]},
  {num: 45, sign: "L", wins:[]}
]

// Write an answer using console.log()

// console.log(players[1].num)

const roundSameSign = (player1: Player, player2: Player):Player => {
  let winner: Player = player1.num < player2.num ? player1 : player2
  winner == player1 ? winner.wins.push(player2.num) :  winner.wins.push(player1.num)
  return winner
};

// let player1 = {num: 420, sign: "L", wins:[]}
// let player2 = {num: 45, sign: "L", wins:[]}

// console.log(roundSameSign(player1, player2));

const round = (player1: Player, player2: Player) => {
  let sign1: Sign = player1.sign;
  let sign2: Sign = player2.sign;
  let winner: Player = player1
  switch (sign1) {
    case 'R':
      winner = ["P", "S"].includes(sign2) ? player2 : player1;
      break;
    case 'P':
      winner = ["C", "L"].includes(sign2) ? player2 : player1;
      break;
    case 'C':
      winner = ["R", "S"].includes(sign2) ? player2 : player1;
      break;
    case 'L':
      winner = ["R", "C"].includes(sign2) ? player2 : player1;
      break;
    case 'S':
      winner = ["L", "P"].includes(sign2) ? player2 : player1;
      break;
    default:
      console.log('Sorry, something went wrong');
  }
  winner == player1 ? winner.wins.push(player2.num) : winner.wins.push(player1.num)
  return winner
}

// let player1 = {num: 420, sign: "L", wins:[]}
// let player2 = {num: 229, sign: "C", wins:[]}
// console.log(roundSameSign(player1, player2));

while (players.length > 1) {
  let winners: Player[] = [];
  let winner: Player
  let l: number = players.length
  for (let i = 0; i <= (l / 2 - 1); i++) {
    let player1: Player = players[i * 2];
    let player2: Player = players[i * 2 + 1];
    if (player1.sign == player2.sign) {
      winner = roundSameSign(player1, player2)
    } else {
      winner = round(player1, player2)
    }
    winners.push(winner)
    // console.log(winner)
  }
  players = winners
}

console.log(players[0].num)
console.log(players[0].wins.join(' '))
